### PR TITLE
Beavertails Finetuning Notebook Changes

### PR DESCRIPTION
**Pull request by aravindputrevu**
More details from the blog PR - https://github.com/cleanlab/web/pull/712

### CODE REVIEW

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 1]
```diff
@@ -0,0 +1,5831 @@
+{
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:1]: can you add something about this to the main Table of Contents in the README?

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 2]
```diff
@@ -0,0 +1,5831 @@
...
+      "metadata": {
+        "id": "MhV4u2N5dfCq"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Role of labelling Data in AI Safety\n",
+        "\n",
+        "In Large Language Model, Data plays a key role as most responses are generated from training data.\n",
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:21]: There are many small issues in the writing throughout. Could you use chatgpt to proofread all your writing? Eg:

Screen Shot 2024-03-09 at 8 19 01 PM

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 3]
```diff
@@ -0,0 +1,5831 @@
...
+      ],
+      "metadata": {
+        "id": "kvkKe8K7Oqq9"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## How to improve Dataset Credibility\n",
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:36]: what is "dataset credibility"? probably want to define something like that before introducing. I would think it means something along the lines of the accuracy of the data, which doesn't make sense in terms of PII / toxicity / sentiment / etc. (none of these are inaccuracies)

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 4]
```diff
@@ -0,0 +1,5831 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Detecting Dataset issues in AI Safety Dataset - Beavertrails\n",
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:6]:
```suggestion
        "## Automatically catching issues in the Beavertrails AI Safety dataset\n",
```

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 5]
```diff
@@ -0,0 +1,5831 @@
...
+        "studio = Studio(API_KEY)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "laHrQ_ImUBdz"
+      },
+      "source": [
+        "Let us upload the dataset to the Cleanlab Studio and creating a project to find issues with the dataset."
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:897]:
```suggestion
        "Let's load the dataset into Cleanlab Studio and creating a project to find issues with the dataset."
```

Comment (jwmueller) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:897]: note the specific language changes being used here and update all of your writing to incorporate them

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 6]
```diff
@@ -0,0 +1,5831 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "cPsLcs-xCBfG"
+      },
+      "source": [
+        "## Using Low Rank Adaptation Technique For Fine-tuning\n",
+        "\n",
+        "Next we setup our LORA finetuning. While giants like Mistral and Phi2 boast impressive capabilities, their computational demands can be hefty. This is where LORA, a lightweight and efficient parameter reduction technique, steps in, offering a compelling alternative.\n",
+        "LORA stands for Low-Rank Optimized Adaptor. It acts as a \"shortcut\" within an LLM, compressing the original model's parameters into a much smaller, low-rank representation. This significantly reduces the model size and computational cost while aiming to maintain similar performance in text generation tasks.\n",
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:1216]: Adapter not Adaptor

Comment (jwmueller) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:1216]: There are many wrong statements in this section. Please remove all statements you are not 100% certain of. For instance, LORA does not reduce a model's size, it just optimizes a small set of adapter parameters that are added to the pretrained model

Comment (jwmueller) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:1216]: QLORA on the other hand does shrink the model by quantization all of its parameters

File: fine_tuning_mistral_beavertails/beavertails.ipynb [hunk 7]
```diff
@@ -0,0 +1,5831 @@
...
+      "source": [
+        "trainer_stats = trainer.train()\n",
+        "train_loss = trainer_stats.training_loss\n",
+        "print(f\"Training loss:{train_loss}\")\n",
+        "\n",
+        "peft_model_id = \"out/\"\n",
+        "trainer.model.save_pretrained(peft_model_id)\n",
+        "tokenizer.save_pretrained(peft_model_id)"
+      ]
+    },
```
Comment (jwmueller) [inline fine_tuning_mistral_beavertails/beavertails.ipynb:1349]: After this you don't want to show how to generate some text from the fine-tuned model? 

Kind of strange to omit that critical part

File: fine_tuning_mistral_beavertails/beavertails.ipynb (hunk 2 above)
Comment (aravindputrevu) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:21]: @jwmueller I ran the notebook through grammarly, much better than the ChatGPT. But as you pointed out, I will redo it again!

File: fine_tuning_mistral_beavertails/beavertails.ipynb (hunk 3 above)
Comment (aravindputrevu) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:36]: I will add a section on "What is Dataset Credibility?". But from the outlook, I see the term as not just accuracy related (i know where you are coming at, when you say accuracy!). Using a dataset to train a model is also about banking on it's reliability, source it has come from, and more importantly in AI safety - trustworthyness. 

Imagine if my fine-tuned LLM gives out profane responses, just like the many other LLM incidents out there.

File: fine_tuning_mistral_beavertails/beavertails.ipynb (hunk 5 above)
Comment (aravindputrevu) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:897]: Thank you, this is helpful!

File: fine_tuning_mistral_beavertails/beavertails.ipynb (hunk 6 above)
Comment (aravindputrevu) [reply fine_tuning_mistral_beavertails/beavertails.ipynb:1216]: I meant it freezes a layer of parameters, but will update the language.

File: README.md [hunk 8]
```diff
@@ -6,3 +6,4 @@ Miscellaneous code made available for purposes of education, reproducibility, an
 | 1 | [few_shot_prompt_selection](few_shot_prompt_selection/few_shot_prompt_selection.ipynb) | Notebook showing how to clean few-shot examples pool to improve prompt template for OpenAI LLM.|
 | 2 | [generate_llm_response](generate_llm_response/generate_llm_response.ipynb) | Notebook showing how to generate LLM responses for customer service requests using Llama 2 and OpenAI's API.|
 | 3 | [fine_tuning_data_curation](fine_tuning_data_curation/fine_tuning_data_curation.ipynb) | Notebook showing how to use Cleanlab TLM and Cleanlab Studio to detect bad data in instruction tuning LLM datasets.|
+| 4 | [fine_tuning_mistral_beavertails](fine_tuning_mistral_beavertails/beavertails.ipynb) | Notebook showing how to detect and analyze human annotated toxicity labels using Cleanlab Studio and thereby generating safer or better responses from LLM.|
```
Comment (jwmueller) [inline README.md:9]:
```suggestion
| 4 | [fine_tuning_mistral_beavertails](fine_tuning_mistral_beavertails/beavertails.ipynb) | Analyze human annotated AI-safety-related labels (like toxicity) using Cleanlab Studio, and thus generate safer responses from LLMs.|
```

**Review by jwmueller — COMMENTED** [review]
LGTM thanks!
